File: Problem_3.cpp
353. Design Snake Game



class SnakeGame {
public:
    SnakeGame(int width, int height, vector<vector<int>>& food) : 
        food_(&food), width_(width), height_(height)  
    {
        snake_.emplace_back(make_pair(0,0));
        pos_[0].emplace(0);
    }
    
    int move(string direction) {
        Coor next = step(direction, snake_.back());

        if (!valid(next)) return -1;
        if (!ProcessFood(next)) 
        {
            Coor tail = snake_.front();
            pos_[tail.first].erase(tail.second);
            snake_.pop_front();
        }
        if (pos_[next.first].count(next.second) > 0) return -1;
        pos_[next.first].emplace(next.second);
        snake_.emplace_back(next);
        return snake_.size() - 1;
    }
private:
    using Coor = pair<int, int>;
    bool valid(Coor coor)
    {
        const bool lower_bounded = (coor.first >= 0 && coor.second >= 0);
        const bool upper_bounded = (coor.first < height_ && coor.second < width_);
        return lower_bounded && upper_bounded;
    }

    Coor step(string dir, Coor coor)
    {
        Coor next = coor;
        if (dir == "U") next.first--;
        else if (dir == "R") next.second++;
        else if (dir == "D") next.first++;
        else if (dir == "L") next.second--;
        return next;
    }

    // returns true if food was eaten, otherwise false
    bool ProcessFood(Coor coor)
    {
        if (food_->size() == food_i_) return false;

        const vector<int>& f = food_->at(food_i_);
        int res = false;
        if (f[0] == coor.first && f[1] == coor.second)
        {
            res = true;
            food_i_++;
        }
        return res;
    }

    int food_i_ = 0;
    vector<vector<int>>* food_{};
    list<Coor> snake_{};
    unordered_map<int, unordered_set<int>> pos_{}; // y : x
    int width_{};
    int height_{};
};

/**
 * Your SnakeGame object will be instantiated and called as such:
 * SnakeGame* obj = new SnakeGame(width, height, food);
 * int param_1 = obj->move(direction);
 */

File: Problem3.py
class SnakeGame:

    def __init__(self, width, height, food):
        '''
        Time complexity - O(1)
        Space complexity - O(n)
        '''
        # Initialize the game attributes
        self.width = width
        self.height = height
        self.food = food
        # Initialize the snake's path with the head at [0, 0]
        self.path = [[0, 0]]
        self.score = 0  # Initialize the initial score
        self.direction_map = {'D': [1, 0],
                              'U': [-1, 0], 'L': [0, -1], 'R': [0, 1]}

    def is_valid_move(self, a, b):
        # Check if the move is valid within the game board
        return a >= 0 and a < self.height and b >= 0 and b < self.width \
            and (self.path[0][0] == a and self.path[0][1] == b or [a, b] not in self.path)

    def move(self, direction):
        # Get the direction's movement mapping
        map_ = self.direction_map[direction]
        head_pos = self.path[-1]  # Get the current head position of the snake
        i, j = head_pos[0] + map_[0], head_pos[1] + \
            map_[1]  # Calculate the new head position

        # Check if the new head position is valid
        if self.is_valid_move(i, j):
            # Check if there is food at the new position
            food = [] if not self.food else self.food[0]
            if food and food[0] == i and food[1] == j:
                self.score += 1  # Increase score and remove the eaten food
                self.food.pop(0)
            else:
                tail = self.path.pop(0)  # Remove the tail if no food was eaten
            # Append the new head position to the path
            self.path.append([i, j])
        else:
            return -1  # Return -1 if the move is invalid

        return self.score  # Return the updated score
